pallets/portfolio/src/lib.rs

```diff
@@ -632,6 +659,67 @@ impl<T: Config> Module<T> {
             Ok(())
         })
     }
+
+    /// Verifies if the portfolios are different, if the move is between the same identity, if the receiving portfolio exists,
+    /// and if the user has access to both portfolios.
+    fn ensure_portfolios_validity_and_permissions(
+        origin: T::Origin,
+        from: PortfolioId,
+        to: PortfolioId,
+    ) -> Result<IdentityId, DispatchError> {
+        let origin_data = Identity::<T>::ensure_origin_call_permissions(origin)?;
+        // Ensures the source and destination portfolios are in fact different
+        ensure!(from != to, Error::<T>::DestinationIsSamePortfolio);
+        // Ensures the source and destination DID are in fact the same
+        ensure!(from.did == to.did, Error::<T>::DifferentIdentityPortfolios);
+        // Ensures the receiving portfolio exists
+        Self::ensure_portfolio_validity(&to)?;
+
+        // Ensures the sender is the custodian and that the secondary key has access to the portfolio.
+        Self::ensure_portfolio_custody_and_permission(
+            from,
+            origin_data.primary_did,
+            origin_data.secondary_key.as_ref(),
+        )?;
+
+        // Ensures the secondary key has access to the receiver's portfolio.
+        Self::ensure_user_portfolio_permission(origin_data.secondary_key.as_ref(), to)?;
+        Ok(origin_data.primary_did)
+    }
+
+    /// Verifies if the sending portfolio has the right balance for the transfer.
+    fn ensure_valid_balances(
+        sender_portfolio: &PortfolioId,
+        items: &[MovePortfolioItem],
+    ) -> DispatchResult {
+        let mut unique_tickers = BTreeSet::new();
+        // Ensure there are sufficient funds for all moves.
+        for item in items {
+            ensure!(
+                unique_tickers.insert(item.ticker),
+                Error::<T>::NoDuplicateAssetsAllowed
+            );
+            Self::ensure_sufficient_balance(sender_portfolio, &item.ticker, item.amount)?;
+        }
+        Ok(())
+    }
+
+    /// Verifies if the sender has the nfts for the transfer and if they are not locked.
+    fn ensure_valid_nfts(sender_portfolio: &PortfolioId, items: &[NFTs]) -> DispatchResult {
```

Comment:
Should we also check that there aren't duplicated NFTs in the `items`?

There was a bug for the fungible asset case where one asset could appear multiple times in `items` and each individual item was valid (i.e. user had sufficient balance) but in aggregate invalid (i.e. user did not have sufficient balance if totals were taken in aggregate across all `items`).